evm/src/cpu/kernel/asm/core/precompiles/expmod.asm

```diff
@@ -1,3 +1,408 @@
+    SWAP1
+    // stack: len, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+
+    DUP1
+    %mul_const(4)
+    // stack: s1=4*len, len, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+    SWAP1
+    // stack: len, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+
+    DUP1
+    %mul_const(3)
+    // stack: out=3*len, len, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+    SWAP1
+    // stack: len, out, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+
+    DUP1
+    %mul_const(2)
+    // stack: m_loc=2*len, len, out, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+    SWAP1
+    // stack: len, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+
+    PUSH 0
+    // stack: b_loc=0, e_loc=len, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+    DUP2
+    // stack: len, b_loc, e_loc, m_loc, out, s1, s2, s3, s4, s5, expmod_contd, len, l_M, kexit_info
+
+    %jump(modexp_bignum)
+
+expmod_contd:
+    // stack: len, l_M, kexit_info
+
+    // Copy the result value from kernel general memory to the parent's return data.
+
+    // Store return data size: l_M (number of bytes).
+    SWAP1
+    // stack: l_M, len, kexit_info
+    %mstore_parent_context_metadata(@CTX_METADATA_RETURNDATA_SIZE)
+    // stack: len, kexit_info
+    DUP1
+    // stack: len, len, kexit_info
+    %mul_const(3)
+    // stack: out=3*len, len, kexit_info
+    DUP2
+    DUP2
+    // stack: out, len, out, len, kexit_info
+    ADD
+    %decrement
+    SWAP1
+    %decrement
+    SWAP1
+    // stack: cur_address=out+len-1, end_address=out-1, len, kexit_info
+    PUSH 0
+    // stack: i=0, cur_address, end_address, len, kexit_info
+
+    // Store in big-endian format.
+expmod_store_loop:
+    // stack: i, cur_address, end_address, len, kexit_info
+    DUP2
+    // stack: cur_address, i, cur_address, end_address, len, kexit_info
+    %mload_kernel_general
+    // stack: cur_limb, i, cur_address, end_address, len, kexit_info
+    DUP2
+    // stack: i, cur_limb, i, cur_address, end_address, len, kexit_info
+    %mul_const(16)
+    // stack: offset=16*i, cur_limb, i, cur_address, end_address, len, kexit_info
+    %stack (offset, cur_limb) -> (@SEGMENT_RETURNDATA, offset, cur_limb, 16)
+    // stack: @SEGMENT_RETURNDATA, offset, cur_limb, 16, i, cur_address, end_address, len, kexit_info
+    %mload_context_metadata(@CTX_METADATA_PARENT_CONTEXT)
+    // stack: parent_ctx, @SEGMENT_RETURNDATA, offset, cur_limb, 16, i, cur_address, end_address, len, kexit_info
+    %mstore_unpacking
+    // stack: offset', i, cur_address, end_address, len, kexit_info
+    POP
+    // stack: i, cur_address, end_address, len, kexit_info
+    %increment
+    SWAP1
+    %decrement
+    SWAP1
+    // stack: i+1, cur_address-1, end_address, len, kexit_info
+    DUP3
+    DUP2
+    EQ
+    ISZERO
+    %jumpi(expmod_store_loop)
+expmod_store_end:
+    // stack: i, cur_address, len, kexit_info
+    %pop3
```

Comment:
This stack comment is incorrect, it should be `// stack: i, cur_address, end_address, len, kexit_info`. And so we should use `%pop4`.